enterprise/server/cmd/goinit/main.go

```diff
@@ -452,6 +471,81 @@ func main() {
 	syscall.Reboot(syscall.LINUX_REBOOT_CMD_RESTART)
 }
 
+// Resizes the rootfs ext4 filesystem to fill the available space on the block
+// device.
+func resizeExt4FS(mountPath string, size int64) error {
+	s := &syscall.Statfs_t{}
+	if err := syscall.Statfs(mountPath, s); err != nil {
+		return status.InternalErrorf("statfs %s: %s", mountPath, err)
+	}
+	blocks := size / s.Bsize
+	fd, err := syscall.Open(mountPath, syscall.O_RDONLY, 0)
+	if err != nil {
+		return err
+	}
+	defer syscall.Close(fd)
+	_, _, errno := syscall.Syscall(
+		syscall.SYS_IOCTL,
+		uintptr(fd),
+		uintptr(EXT4_IOC_RESIZE_FS),
+		uintptr(unsafe.Pointer(&blocks)),
+	)
+	if errno != 0 {
+		return status.InternalErrorf("EXT4_IOC_RESIZE_FS: errno %s", errno)
+	}
+	return nil
+}
+
+// startRootfsClientProcess starts a background process that acts as the NBD
+// client for the rootfs. It creates the rootfs NBD device and handles all IO
+// requests to it. This function returns when the rootfs is mounted or the
+// rootfs
+func startRootfsClientProcess() error {
+	cmd := exec.Command(os.Args[0], append(os.Args[1:], "--rootfs_client")...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	errCh := make(chan error)
+	go func() { errCh <- cmd.Wait() }()
+	for {
+		// TODO(bduffany): Do something better than file polling here.
```

Comment:
I see it's a TODO so no pressure to do it now, but one potential implementation is using unix.Poll on a file descriptor like we do [here](https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/server/remote_execution/uffd/uffd.go#L227)